### DISCUSSION TITLE
[Implementation Question]: Should we override Equals and ToString?

### DISCUSSION THREAD

**Discussion by GoEddie** [dotnet/spark#585] [Ideas]
Does it make sense to override these in spark objects and pass them back to the JVM (or provide our own implementation)? 

Is there a preference for overriding or not overriding? I did a quick search and could see a few "override ToString" so wasn't sure what to do.

**Comment by imback82** [comment]
Could you elaborate on the use case?

**Comment by GoEddie** [comment]
In spark toString is overridden (not sure if it is always or just sometimes) so doing:

https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/Column.scala#L153

```scala
col("a").toString
```
returns a string "a"

if you do that on a Col in .NET you get:

```csharp
Col("a").ToString()
```

returns a string "Microsoft.Spark.Sql.Column".

It feels like if we forwarded ToString and Equals then we will be more consistent with whatever spark does (or doesn't if it doesn't override them in some cases).

**Comment by imback82** [comment]
Got it. Yea, I think we can forward them to Scala side.

**Comment by GoEddie** [comment]
Thanks, Will do

**Comment by imback82** [comment]
I am curious what the notebook experience will be like since it depends on `ToString`. @suhsteve any concern?

**Comment by suhsteve** [comment]
There may be some unintended side effects by override Equals ?  One example may be if a user is adding/checking these jvm objects against C# dictionary/hashset, any collisions would create jvm calls.

**Comment by suhsteve** [comment]
> I am curious what the notebook experience will be like since it depends on `ToString`. @suhsteve any concern?

I don't see anything too concerning with toString with notebooks atm.

**Comment by imback82** [comment]
I am thinking we should support `Equals` and `ToString` for both C# side and JVM side (C# as the first citizen and a variation of `Equals` and `ToString` for the Scala). What do you guys think?

**Comment by GoEddie** [comment]
I’d say pass toString on, this is another example: https://github.com/dotnet/spark/issues/594

**Comment by imback82** [comment]
Yea, we can start with `ToString`. Should we pass thru for all the wrapper classes or the ones that override `toString` on the Scala side? (I see that we have few classes that override `ToString` to call the Scala side b/c it's overridden on the Scala side).
